File: hikyuu/gui/shmserver.py
#!/usr/bin/python
# -*- coding: utf8 -*-
#
# Create on: 2026-09-07
#    Author: fasiondog

import time

import click

import hikyuu as hku


@click.command()
@click.option('--datadir', default="", help='Data directory; when empty, the [hikyuu] datadir in hikyuu.ini is used')
@click.option('--publish_shm', default=True, type=bool, help='Whether to publish the shared memory snapshot (K-line hot data + basic info)')
@click.option('--recv_spot', default=True, type=bool, help='Whether this process receives realtime quotes and mirrors them to the tail of the snapshot')
@click.option('--config', 'config_file', default="", help='The hikyuu configuration file path; when empty, the default ~/.hikyuu/hikyuu.ini is used')
def main(datadir, publish_shm, recv_spot, config_file):
    """在当前进程内启动 shm(共享内存)数据服务（独立 VIP 插件，需有效授权），常驻供其他 hikyuu 进程零拷贝读取。

    服务不会自动产生：需在本进程显式启动；其他进程显式开启 use_shm_server=True 且同一
    datadir 已有服务时，将自动作为客户端接入。按 Ctrl-C 停止服务。
    """
    # 服务端进程只作发布者、不接入其他 shm 服务：显式关闭客户端探测（use_shm_server=False）。
    # 该开关默认即关闭，此处仍显式声明以防用户配置文件中已开启；否则 init 阶段会先当客户端
    # 探测既有服务、白等约 10s 后打印 "fallback to standalone mode" 告警。
    # 同时确保按 --config（或默认）配置完成数据初始化（import hikyuu 本身不加载数据）。
    load_kwargs = {"use_shm_server": False}
    if config_file:
        load_kwargs["config_file"] = config_file
    hku.load_hikyuu(**load_kwargs)

    try:
        if not hku.start_shm_server(datadir, publish_shm=publish_shm, recv_spot=recv_spot):
            click.echo("start_shm_server returned False: missing plug-in / invalid license or data not ready; this process did not become the server.", err=True)
            return

        click.echo("The shm data service has started, press Ctrl-C to stop.")
        while True:
            try:
                time.sleep(1)
            except KeyboardInterrupt:
                break

    finally:
        hku.stop_shm_server()


if __name__ == "__main__":
    main()

File: hikyuu_pywrap/plugin/_shmserver.cpp
/*
 *  Copyright (c) 2026 hikyuu.org
 *
 *  Created on: 2026-09-06
 *      Author: fasiondog
 */

#include <hikyuu/plugin/shmserver.h>
#include "../pybind_utils.h"

namespace py = pybind11;
using namespace hku;

void export_plugin_shmserver(py::module& m) {
    m.def(
      "start_shm_server", startShmServer, py::arg("datadir") = std::string(),
      py::arg("publish_shm") = true, py::arg("recv_spot") = true,
      R"(start_shm_server(datadir: str='', publish_shm: bool=True, recv_spot: bool=True) -> bool

    Start the shm (shared memory) data server within the current process, for the other hikyuu processes to read with zero copy as the clients, avoiding the repeated data loading of the multiple processes.
    The service is provided by the standalone shmserver plugin (a VIP plugin, requiring a valid license).
    The client processes need to explicitly enable use_shm_server (in the configuration file or load_hikyuu(use_shm_server=True)) to join this service.

    It must be called after the hikyuu initialization (import hikyuu completes the initialization by default); calling it before the initialization will return False because the data is not ready.

    :param str datadir: the data directory; when it is empty, the current StockManager data directory is used
    :param bool publish_shm: whether to publish the two kinds of the shared memory snapshots (the K-line hot data + the basic information)
    :param bool recv_spot: whether this process receives the real-time market data (internally subscribing to quotation_server and driving the real-time updates)
    :return: return True when started successfully; return False when this process is already in the client mode, the plugin is missing or the license is invalid)");

    m.def("stop_shm_server", stopShmServer, R"(stop_shm_server() -> None

    Stop the shm data server within the current process, releasing the shared memory segments and unregistering the related hooks)");

    m.def("is_shm_server_running", isShmServerRunning, R"(is_shm_server_running() -> bool

    Query whether the shm data server within the current process is running

    :return: return True when running, otherwise False)");
}
